### PR TITLE
BUG: fix calling Table.group_by on an empty table

### LINKED ISSUE #11884: group_by() on a zero-length Table fails
### Description

Using `group_by()` on a zero-length Table generates an exception.

### Expected behavior

I expect a zero-length table as output.

### Actual behavior

I get an IndexError.

### Steps to Reproduce

```python
>>> t = Table(names=['a', 'b'])                                                                                                  

>>> t                                                                                                                            
<Table length=0>
   a       b   
float64 float64
------- -------

>>> t.group_by('a')                                                                                                              
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-6-d83787522827> in <module>
----> 1 t.group_by('a')

~/miniconda3/envs/ska3/lib/python3.8/site-packages/astropy/table/table.py in group_by(self, keys)
   3291             New table with groups set
   3292         """
-> 3293         return groups.table_group_by(self, keys)
   3294 
   3295     def to_pandas(self, index=None, use_nullable_int=True):

~/miniconda3/envs/ska3/lib/python3.8/site-packages/astropy/table/groups.py in table_group_by(table, keys)
     16     # index copies are unnecessary and slow down _table_group_by
     17     with table.index_mode('discard_on_copy'):
---> 18         return _table_group_by(table, keys)
     19 
     20 

~/miniconda3/envs/ska3/lib/python3.8/site-packages/astropy/table/groups.py in _table_group_by(table, keys)
    103     # Take the subset of the original keys at the indices values (group boundaries).
    104     out = table.__class__(table[idx_sort])
--> 105     out_keys = table_keys[indices[:-1]]
    106     if isinstance(out_keys, Table):
    107         out_keys.meta['grouped_by_table_cols'] = grouped_by_table_cols

~/miniconda3/envs/ska3/lib/python3.8/site-packages/astropy/table/table.py in __getitem__(self, item)
   1662             # is produced by np.where, as in t[np.where(t['a'] > 2)]
   1663             # For all, a new table is constructed with slice of all columns
-> 1664             return self._new_from_slice(item)
   1665         else:
   1666             raise ValueError('Illegal type {} for table item access'

~/miniconda3/envs/ska3/lib/python3.8/site-packages/astropy/table/table.py in _new_from_slice(self, slice_)
   1179         newcols = []
   1180         for col in self.columns.values():
-> 1181             newcol = col[slice_]
   1182 
   1183             # Note in line below, use direct attribute access to col.indices for Column

astropy/table/_column_mixins.pyx in astropy.table._column_mixins._ColumnGetitemShim.__getitem__()

astropy/table/_column_mixins.pyx in astropy.table._column_mixins.base_getitem()

astropy/table/_column_mixins.pyx in astropy.table._column_mixins.column_getitem()

IndexError: index 0 is out of bounds for axis 0 with size 0


```

### System Details

```
macOS-10.16-x86_64-i386-64bit
Python 3.8.3 (default, Jul  2 2020, 11:26:31) 
[Clang 10.0.0 ]
Numpy 1.18.5
astropy 4.2
Scipy 1.5.0
Matplotlib 3.2.2
```

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Fixes #11884

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/table/tests/test_table.py [hunk 1]
```diff
@@ -3178,6 +3178,12 @@ def test_rows_with_mixins():
     t.group_by("obs")
 
 
+def test_group_by_empty_table():
+    # see https://github.com/astropy/astropy/issues/11884
+    t = Table(names=["a", "b"])
+    t.group_by("a")
```
Comment (pllim) [inline astropy/table/tests/test_table.py:3184]: Do we want to check the state of `t` after this or it doesn't matter?

File: docs/changes/table/15938.bugfix.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix calling ``Table.group_by`` on an empty table.
```
Comment (pllim) [inline docs/changes/table/15938.bugfix.rst:1]:
```suggestion
Calling ``Table.group_by`` on an empty table no longer raises an exception.
```

**Review by pllim — COMMENTED** [review]
Thanks!

File: astropy/table/tests/test_table.py (hunk 1 above)
Comment (neutrinoceros) [reply astropy/table/tests/test_table.py:3184]: I think we should, but I don't have a clear idea what to check for.

Comment (pllim) [reply astropy/table/tests/test_table.py:3184]: OK hopefully @taldcroft can advise.

Comment (neutrinoceros) [reply astropy/table/tests/test_table.py:3184]: Let's remind @taldcroft about this :)

**Comment by taldcroft** [comment]
@neutrinoceros - thanks for this PR but I think I have a solution that more narrowly addresses the original issue: https://github.com/astropy/astropy/issues/11884#issuecomment-1959567246.

The change you made to `Table.__getitem__` would introduce a new bug:
```
In [2]: t = Table(names=["a", "b"])

In [3]: t[np.array([10])]  # Should raise IndexError
Out[3]: 
<Table length=0>
   a       b   
float64 float64
------- -------
```

**Comment by neutrinoceros** [comment]
No worries, thanks for noticing !

**Pull request closed without merging**